libs/wire-subsystems/src/Wire/EnterpriseLoginSubsystem/Interpreter.hs

```diff
@@ -0,0 +1,321 @@
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  DomainRegistrationUpdate ->
+  Sem r ()
+updateDomainRegistrationImpl domain update = do
+  validate update
+  mOld <- tryGetDomainRegistrationImpl domain
+  let url =
+        "PUT /i/domain-registration/"
+          <> fromText (domainText domain)
+  new <- case mOld of
+    Just dr -> do
+      let new = dr {teamInvite = update.teamInvite, domainRedirect = update.domainRedirect} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Nothing -> do
+      let dr = DomainRegistration domain update.domainRedirect update.teamInvite Nothing
+      upsert $ toStored dr
+      pure dr
+  sendAuditMail url "Domain registration updated" mOld (Just new)
+
+lockDomainImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+lockDomainImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  let new = DomainRegistration domain Locked Allowed Nothing
+  upsert $ toStored new
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/lock"
+  sendAuditMail url "Domain locked" mOld (Just new)
+
+unlockDomainImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+unlockDomainImpl domain = do
+  old <- getDomainRegistrationImpl domain
+  let new = old {domainRedirect = None} :: DomainRegistration
+  case old.domainRedirect of
+    Locked -> upsert $ toStored new
+    _ -> throw EnterpriseLoginSubsystemUnlockError
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/unlock"
+  sendAuditMail url "Domain locked" (Just old) (Just new)
+
+preAuthorizeImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+preAuthorizeImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  new <- case mOld of
+    Nothing -> do
+      let new = DomainRegistration domain PreAuthorized Allowed Nothing
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == None -> do
+      let new = sdr {domainRedirect = PreAuthorized} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == PreAuthorized -> pure sdr
+    _ -> throw EnterpriseLoginSubsystemErrorInvalidDomainRedirect
```

Comment:
this could use a more helpful explanation as a string arg to `EnterpriseLoginSubsystemErrorInvalidDomainRedirect`.  what about "could not pre-authorize domain; current domain settings: <show>"?

also, why not `old` instead of `sdr`?